dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt

```diff
@@ -78,13 +79,50 @@ private class SymbolSampleAnalysisEnvironment(
         return SampleSnippet(imports, body)
     }
 
+    // copy-pasted from StdLib 1.5
+    // TODO use it from Stdlib
+     private inline fun <T, R : Any> Iterable<T>.firstNotNullOfOrNull(transform: (T) -> R?): R? {
+        for (element in this) {
+            val result = transform(element)
+            if (result != null) {
+                return result
+            }
+        }
+        return null
+    }
+
+    // TODO: remove after KT-53669 and use [org.jetbrains.kotlin.analysis.api.symbols.sourcePsiSafe] from Analysis API
+    private inline fun <reified PSI : PsiElement> KtSymbol.kotlinAndJavaSourcePsiSafe(): PSI? {
+        // TODO: support Java sources after KT-53669
+        val sourcePsi = when (origin) {
+            KtSymbolOrigin.SOURCE -> this.psi
+            KtSymbolOrigin.JAVA -> this.psi
+
+            KtSymbolOrigin.SOURCE_MEMBER_GENERATED -> null
+            KtSymbolOrigin.LIBRARY -> null
+            KtSymbolOrigin.SAM_CONSTRUCTOR -> null
+            KtSymbolOrigin.INTERSECTION_OVERRIDE -> null
+            KtSymbolOrigin.SUBSTITUTION_OVERRIDE -> null
+            KtSymbolOrigin.DELEGATED -> null
+            KtSymbolOrigin.JAVA_SYNTHETIC_PROPERTY -> null
+            KtSymbolOrigin.PROPERTY_BACKING_FIELD -> null
+            KtSymbolOrigin.PLUGIN -> null
+            KtSymbolOrigin.JS_DYNAMIC -> null
+        }
+
+        return sourcePsi as? PSI
+    }
+
     private fun findPsiElement(sourceSet: DokkaSourceSet, fqLink: String): PsiElement? {
-        val ktSourceModule = samplesKotlinAnalysis.getModuleOrNull(sourceSet)
-            ?: projectKotlinAnalysis.getModule(sourceSet)
+        // fallback to default roots of the source set even if sample roots are assigned
+        val ktSourceModules =
+            listOfNotNull(samplesKotlinAnalysis.getModuleOrNull(sourceSet), projectKotlinAnalysis.getModule(sourceSet))
 
-        return analyze(ktSourceModule) {
-            resolveKDocTextLinkSymbol(fqLink)
-                ?.sourcePsiSafe()
+        return ktSourceModules.firstNotNullOfOrNull { ktSourceModule ->
```

Comment:
Calling `listOfNotNull` to then do `list.firstNotNullOfOrNull` seems excessive in this case, and could be avoided like so

```kotlin
    private fun findPsiElement(sourceSet: DokkaSourceSet, fqLink: String): PsiElement? {
        // fallback to default roots of the source set even if sample roots are assigned,
        // because `@sample` tag can contain links to local project functions
        return samplesKotlinAnalysis.findPsiElement(sourceSet, fqLink)
            ?: projectKotlinAnalysis.findPsiElement(sourceSet, fqLink)
    }

    private fun KotlinAnalysis.findPsiElement(sourceSet: DokkaSourceSet, fqLink: String): PsiElement? {
        val ktSourceModule = this.getModuleOrNull(sourceSet) ?: return null
        return analyze(ktSourceModule) {
            resolveKDocTextLinkSymbol(fqLink)
                ?.kotlinAndJavaSourcePsiSafe()
        }
    }
```